src/ai/backend/test/templates/session/utils.py

```diff
@@ -0,0 +1,136 @@
+import asyncio
+from typing import Awaitable, Callable
+from uuid import UUID
+
+from ai.backend.client.func.session import ComputeSession
+from ai.backend.client.session import AsyncSession
+from ai.backend.common.json import load_json
+
+
+async def verify_batch_session(
+    client_session: AsyncSession,
+    session_name: str,
+    timeout: float,
+    session_factory: Callable[[], Awaitable["ComputeSession"]],
+) -> UUID:
+    EXPECTED_EVENTS: set[str] = {
+        "session_enqueued",
+        "session_scheduled",
+        "kernel_preparing",
+        "kernel_creating",
+        "kernel_started",
+        "session_started",
+        "session_terminating",
+        "session_terminated",
+        "kernel_terminating",
+        "kernel_terminated",
+        "session_success",
+    }
+
+    collected_events: set[str] = set()
+
+    async def collect_events() -> None:
+        async with client_session.ComputeSession(session_name).listen_events() as events:
+            async for ev in events:
+                collected_events.add(ev.event)
+                if ev.event == "session_failure" or ev.event == "session_cancelled":
+                    raise RuntimeError(f"BatchSession failed with event: {ev.event}")
+                if collected_events == EXPECTED_EVENTS:
+                    break
+
+    listener = asyncio.create_task(asyncio.wait_for(collect_events(), timeout))
+
+    created = await session_factory()
+    assert created.created, "Session should be created successfully"
+    assert created.name == session_name, "Session name mismatch"
+
+    try:
+        await listener
+        assert created.status in {"TERMINATING", "TERMINATED"}, (
+            f"Unexpected final status: {created.status}"
+        )
+        if created.id is None:
+            raise RuntimeError("Session ID is None after creation")
+        return created.id
+    except asyncio.TimeoutError as e:
+        raise asyncio.TimeoutError(
+            f"Timed out after {timeout}s; events so far: {collected_events}"
+        ) from e
+
+
+async def verify_interactive_session_creation(
+    client_session: AsyncSession,
+    session_name: str,
+    timeout: float,
+    create_session: Callable[[], Awaitable["ComputeSession"]],
+) -> UUID:
+    EXPECTED_CREATION_EVENTS: set[str] = {
+        "session_enqueued",
+        "session_scheduled",
+        "kernel_preparing",
+        "kernel_creating",
+        "kernel_started",
+        "session_started",
+    }
+
+    collected: set[str] = set()
+
+    async def collect_events() -> None:
+        async with client_session.ComputeSession(session_name).listen_events() as evs:
+            async for ev in evs:
+                collected.add(ev.event)
+                if ev.event == "session_cancelled":
+                    raise RuntimeError("Session creation was cancelled")
```

Comment:
Wouldn't it be better to raise an error when an event other than the expected event occurs, rather than adding exception handling for specific events like this?

Reply:
It seems like it would also be easier to generalize that way.